File: 0242-valid-anagram/0242-valid-anagram.java
class Solution {
    public boolean isAnagram(String s, String t) {
        HashMap <Character, Integer> store = new HashMap<>();
        if (s.length() != t.length()) return false;
        
        for (char k : s.toCharArray()) {
            store.put(k, store.getOrDefault(k, 0) + 1);
        }

        for (char k : t.toCharArray()) {
            System.out.println(store.get(k));
            if (!store.containsKey(k) || store.get(k) == 0) {
                return false;
            } else {
                store.put(k, store.get(k) - 1);
            }
        }
        // for (char k : store.keySet()) {
        //     System.out.println("key " + k + " value " + store.get(k));
        // }

        return true;
    }
}

File: 0242-valid-anagram/0242-valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # brute: sort and compare if they are the same O(nlogn + n)
        # optimal: hashmap of one and iterate through the second subtracting values O(2n) O(n)
        if len(s) != len(t):
            return False
        store = {}
        for char in s:
            if char in store:
                store[char] += 1
            else:
                store[char] = 1
        print(store)
        for char in t:
            if char in store:
                store[char] -= 1
            if not (char in store) or store[char] < 0:
                return False
        return True
